### PR TITLE
Add RequestLoggingTransformer mechanism

### PR DESCRIPTION
**Pull request by ean5533**
This will allow services to inject custom transformers that will get applied right before request/response bodies are logged. The main motivating use case is wanting to trim out noisy data from response bodies.

### CODE REVIEW

File: misk/src/test/kotlin/misk/web/interceptors/RequestLoggingInterceptorTest.kt [hunk 1]
```diff
@@ -354,3 +398,31 @@ internal class ConfigOverrideAction @Inject constructor() : WebAction {
   ) // these values overridden by RequestLoggingConfig binding above
   fun call(@PathParam message: String): String = "echo: $message"
 }
+
+/**
+ * A [RequestLoggingTransformer] that doesn't like seeing "Evan" in String response bodies
+ */
+internal class EvanHatingTransformer @Inject constructor() : RequestLoggingTransformer {
```
Comment (keeferrourke) [inline misk/src/test/kotlin/misk/web/interceptors/RequestLoggingInterceptorTest.kt:405]: I disagree with the premise of this transformer 😂 😭

Comment (ean5533) [reply misk/src/test/kotlin/misk/web/interceptors/RequestLoggingInterceptorTest.kt:405]: I don't condone name-based prejudice but I think it's import for Transformers to be able to express their views.

File: misk/src/main/kotlin/misk/web/interceptors/RequestLoggingInterceptor.kt [hunk 2]
```diff
@@ -131,7 +134,11 @@ class RequestLoggingInterceptor internal constructor(
     val sampling = if (isError) errorBodySampling else bodySampling
     val randomDouble = random.current().nextDouble()
     if (randomDouble < sampling) {
-      val requestResponseBody = bodyCapture.get()
+      val requestResponseBody =
+        requestLoggingTransformers.foldRight(bodyCapture.get()) { transformer, body ->
+          transformer.tryTransform(body)
+        }
```
Comment (keeferrourke) [inline misk/src/main/kotlin/misk/web/interceptors/RequestLoggingInterceptor.kt:140]: It'd be nice to see a test for behaviour with multiple transformers applied to the same request. I have a feeling there might be a gotcha with ordering these transformations?

**Comment by kyeotic** [comment]
> Because of this, my current stance is that there is no need to add the complexity of a configureProviders-style binding API -- just stick with vanilla multibind.

I agree

**Review by snigdhabanda — APPROVED** [review]

**Review by adrw — APPROVED** [review]

File: misk/src/test/kotlin/misk/web/interceptors/RequestLoggingInterceptorTest.kt (hunk 1 above)
Comment (adrw) [reply misk/src/test/kotlin/misk/web/interceptors/RequestLoggingInterceptorTest.kt:411]: For future maintainers, I'd lean towards renaming this to something less personal.

File: misk/src/main/kotlin/misk/web/interceptors/RequestLoggingInterceptor.kt [hunk 3]
```diff
@@ -45,6 +46,7 @@ class RequestLoggingInterceptor internal constructor(
     private val logRateLimiter: LogRateLimiter,
     private val deployment: Deployment,
     private val configs: Set<RequestLoggingConfig>,
+    private val requestLoggingTransformers: List<RequestLoggingTransformer>,
```
[the remark this answers is not in the document]
Comment (ean5533) [reply misk/src/main/kotlin/misk/web/interceptors/RequestLoggingInterceptor.kt:49]: It works by virtue of: https://github.com/cashapp/misk/blob/4b00d84294aabea58a3bd2a1f24ab1fa1547b45e/misk-inject/src/main/kotlin/misk/inject/KAbstractModule.kt#L67-L88

[the remark this answers is not in the document]
Comment (ean5533) [reply misk/src/main/kotlin/misk/web/interceptors/RequestLoggingInterceptor.kt:49]: We could bring it in as a `Set` just fine; honestly the only reason I brought it in as a `List` was just to make it easier to write the unit test that proved that all transformers run even if one of them throws

[the remark this answers is not in the document]
Comment (cconroy) [reply misk/src/main/kotlin/misk/web/interceptors/RequestLoggingInterceptor.kt:49]: Eh, I don't feel strongly..local consistency is maybe worth going for here but it's NBD

**Pull request merged**